### PR TITLE
Performance optimizations on separate manager preview split, improved cold start, restart & rebuild

### PR DESCRIPTION
**Pull request by ndelangen**
This PR adds optimisation to the preview-manager split.

It uses a couple of techniques to make the cold start and hot restart of webpack faster.

I've done some testing and the performance of the manager seems heavily dependent on how busy the preview is...

I think this means it WOULD be logical to move the manager back out into it's own process.

Overall I'm seeing a 41s start-up on the next branch, this one has a bit more reasonable time of 11s.

When I stop and re-start the dev server I'm seeing a boot up time of 9 seconds.

---

I also added a bunch of logs and profile data from webpack to be able to somewhat debug what is happening.

The total time for manager and preview are now printed in the terminal for end-users. They'll be more incentivised to report issues I hope, or at the very least they'll get feedback if performance is improving or deteriorating.

---

- One of the techniques is a shared cache between the 2 instances of webpack
- One is using a DLL plugin for the entire lib/ui
- One is to resolve the configs in parallel

---

The debug info is written to the cache dir when you perform a smoke-test.

The debug information includes all the stats for both manager and preview. You can dump them into something like:
https://chrisbateman.github.io/webpack-visualizer/ or http://webpack.github.io/analyse/#modules

The lib/ui will be missing because it's essentially pre-bundled.

Another piece of debugging data is the full event log of webpack.
Maybe it's a bug, maybe I'm doing something wrong, but it seems to be the combined log of both webpack instances. either way it can be insightful what is taking webpack so long.

### CODE REVIEW

File: lib/core/src/server/build-dev.js [hunk 1]
```diff
@@ -139,11 +143,14 @@ export async function buildDevStandalone(options) {
...
     const server = getServer(app, options);
 
+    app.use('/sb_dll', express.static(dllPath, { index: false }));
+
     applyStatic(app, options);
 
     const storybookMiddleware = await storybook(options);
 
     app.use(storybookMiddleware);
+    // ;
```
Comment (hipstersmoothie) [inline lib/core/src/server/build-dev.js]: seems like an unecassary line

```suggestion
```

File: lib/core/src/server/build-static.js [hunk 2]
```diff
@@ -72,42 +40,105 @@ export async function buildStaticStandalone(options) {
...
+          webpack(managerConfig).run((err, stats) => {
+            if (err || !stats || stats.hasErrors()) {
+              logger.error('Failed to build the manager');
+              // eslint-disable-next-line no-unused-expressions
+              err && logger.error(err.message);
+              // eslint-disable-next-line no-unused-expressions
+              stats && stats.hasErrors() && stats.toJson().errors.forEach(e => logger.error(e));
+              process.exitCode = 1;
+
+              reject(stats);
```
Comment (hipstersmoothie) [inline lib/core/src/server/build-static.js]: if you don't return here it will bother reject and resolve I'm pretty sure

```suggestion
              return reject(stats);
```

File: lib/core/src/server/build-static.js [hunk 3]
```diff
@@ -72,42 +40,105 @@ export async function buildStaticStandalone(options) {
...
+        new Promise((resolve, reject) => {
+          webpack(config).run((err, stats) => {
+            if (err || !stats || stats.hasErrors()) {
+              logger.error('Failed to build the preview');
+              // eslint-disable-next-line no-unused-expressions
+              err && logger.error(err.message);
+              // eslint-disable-next-line no-unused-expressions
+              stats && stats.hasErrors() && stats.toJson().errors.forEach(e => logger.error(e));
+              process.exitCode = 1;
+              reject(err);
```
Comment (hipstersmoothie) [inline lib/core/src/server/build-static.js]: Same return problem here

File: lib/core/src/server/build-static.js [hunk 4]
```diff
@@ -72,42 +40,105 @@ export async function buildStaticStandalone(options) {
...
-      compiler.watch({}, webpackCb);
-    } else {
-      compiler.run(webpackCb);
-    }
-  });
+    });
 }
 
-export async function buildStatic({ packageJson, ...loadOptions }) {
+export function buildStatic({ packageJson, ...loadOptions }) {
```
Comment (hipstersmoothie) [inline lib/core/src/server/build-static.js]: since this function returns a promise I think it should be async still

File: lib/core/src/server/dev-server.js [hunk 5]
```diff
@@ -18,47 +18,53 @@ export const webpackValid = new Promise((resolve, reject) => {
...
+            }
+          }
+        );
+      })
+  );
 
   const iframeConfig = await loadConfig({
     configType,
     outputDir,
+    // cache,
```
Comment (hipstersmoothie) [inline lib/core/src/server/dev-server.js]: remove comment

File: lib/core/src/server/manager/manager-preset.js [hunk 6]
```diff
@@ -7,7 +7,8 @@ export async function managerWebpack(_, options) {
 
 export async function managerEntries(_, options) {
   const { presets } = options;
-  const entries = [require.resolve('../common/polyfills')];
+  // const entries = [require.resolve('../common/polyfills')];
```
Comment (hipstersmoothie) [inline lib/core/src/server/manager/manager-preset.js]: remove comment

File: lib/ui/scripts/config.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  module: {
+    rules: [
+      {
+        test: /\.jsx?$/,
+        use: [
+          {
+            loader: 'babel-loader',
+            options: {
+              presets: [
+                ['@babel/preset-env', { shippedProposals: true, modules: false }],
```
Comment (hipstersmoothie) [inline lib/ui/scripts/config.js]: I feel like i've seen this above. Maybe the babel rcs could be moved to a common location?

File: lib/ui/scripts/script.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    webpack(c).run((err, stats) => {
+      if (err || stats.hasErrors() || stats.hasWarnings()) {
+        rej(stats);
+      }
+      res(stats);
+    });
+  });
+
+const run = async () => {
+  const out2 = await webpackAsPromised(
```
Comment (hipstersmoothie) [inline lib/ui/scripts/script.js]: this function can just return the promise instead of awaiting for it. the only reason you'd await in this function is if you wanted specific errors in the try/catch

File: lib/core/src/server/build-static.js (hunk 4 above)
Comment (ndelangen) [reply lib/core/src/server/build-static.js]: `async` should only be used in combination with `await` right?

File: lib/ui/scripts/config.js (hunk 7 above)
Comment (ndelangen) [reply lib/ui/scripts/config.js]: Sounds like an improvement/refactor we could do a bit later. Good suggestion!

File: lib/core/src/server/build-static.js (hunk 4 above)
Comment (hipstersmoothie) [reply lib/core/src/server/build-static.js]: The typescript compiler is becoming part of my brain! https://javascript.info/async-await The `async` keyword on a function just means that the function will always return a promise. It doesn't have to be used in combination with `await` at all. 

Probably an unneeded change, but it does communicate easily that this function returns a promise

**Comment by ndelangen** [comment]
Who can give this a final check, so we can move ahead and merge. That way we can get this in a "next" release?

**Review by tmeasday — APPROVED** [review]
Code LGTM.

I compared it to `next` on my iMac:

On `next`:

```
╭──────────────────────────────────────────────────╮
│                                                  │
│   Storybook 4.1.0-alpha.8 started                │
│   26 s for manager and 12 s for preview          │
│                                                  │
│   Local:            http://localhost:9011/       │
│   On your network:  http://192.168.0.22:9011/    │
│                                                  │
╰──────────────────────────────────────────────────╯
```

On this branch:
```
╭──────────────────────────────────────────────────╮
│                                                  │
│   Storybook 4.1.0-alpha.8 started                │
│   7.46 s for manager and 9.1 s for preview       │
│                                                  │
│   Local:            http://localhost:9011/       │
│   On your network:  http://192.168.0.22:9011/    │
│                                                  │
╰──────────────────────────────────────────────────╯
```

Seems positive.

It did output this:
```
Error: ENOENT: no such file or directory, stat '/Users/tom/OSS/storybook/lib/core/dist/public/runtime~main.3aaa912fd37e258bb4b5.bundle.js'
Error: ENOENT: no such file or directory, stat '/Users/tom/OSS/storybook/lib/core/dist/public/main.ef442cdddb07ad8080c3.bundle.js'
Error: ENOENT: no such file or directory, stat '/Users/tom/OSS/storybook/lib/core/dist/public/vendors~main.9171936130daf921c71d.bundle.js'
Error: ENOENT: no such file or directory, stat '/Users/tom/OSS/storybook/lib/core/dist/public/runtime~main.0f4ed394e749803fae17.bundle.js'
Error: ENOENT: no such file or directory, stat '/Users/tom/OSS/storybook/lib/core/dist/public/main.0f4ed394e749803fae17.bundle.js'
Error: ENOENT: no such file or directory, stat '/Users/tom/OSS/storybook/lib/core/dist/public/vendors~main.0f4ed394e749803fae17.bundle.js'
```

It didn't seem to cause any issues though, and didn't appear on a second run.

File: lib/core/src/server/build-static.js [hunk 9]
```diff
@@ -72,42 +40,102 @@ export async function buildStaticStandalone(options) {
...
+        packageJson,
+        corePresets: [require.resolve('./preview/preview-preset.js')],
+        overridePresets: [require.resolve('./preview/custom-webpack-preset.js')],
+        ...options,
+      });
+    })
+    .then(
+      config =>
+        new Promise((resolve, reject) => {
+          webpack(config).run((err, stats) => {
```
Comment (igor-dv) [inline lib/core/src/server/build-static.js]: in case we have a "watch" parameter passed, we should run `webpack(config).watch()` for manager. Otherwise it's a breaking change =(

File: lib/core/src/server/build-static.js [hunk 10]
```diff
@@ -11,54 +11,22 @@ import loadManagerConfig from './manager/manager-config';
...
 
-export async function buildStaticStandalone(options) {
-  const { staticDir, watch, configDir, packageJson } = options;
+export function buildStaticStandalone(options) {
+  const { staticDir, configDir, packageJson } = options;
 
   const configType = 'PRODUCTION';
   const outputDir = path.join(process.cwd(), options.outputDir);
 
+  const dllPath = path.join(__dirname, '../../dll/storybook_ui_dll.js');
```
Comment (igor-dv) [inline lib/core/src/server/build-static.js]: why dlls are located in the core package and not in the cache?

File: lib/core/src/server/manager/manager-webpack.config.js [hunk 11]
```diff
@@ -1,30 +1,40 @@
...
 
+import findCacheDir from 'find-cache-dir';
+
 import { version } from '../../../package.json';
 import { getManagerHeadHtml } from '../utils/template';
 import { loadEnv, getBabelRuntimePath } from '../config/utils';
 
-export default ({ configDir, entries, outputDir }) => {
-  const { raw, stringified } = loadEnv();
+const r = require.resolve('@storybook/core/package.json').replace('package.json', '');
```
Comment (igor-dv) [inline lib/core/src/server/manager/manager-webpack.config.js]:
```suggestion
const coreDirName = path.dirname(require.resolve('@storybook/core/package.json'));
```

File: lib/core/src/server/manager/manager-webpack.config.js [hunk 12]
```diff
@@ -1,30 +1,40 @@
+import path from 'path';
```
Comment (igor-dv) [inline lib/core/src/server/manager/manager-webpack.config.js]: `@ndelangen/html-webpack-harddisk-plugin` should be removed from pkg.json ?

File: lib/ui/scripts/config.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
+import path from 'path';
+import webpack from 'webpack';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import UglifyJsPlugin from 'uglifyjs-webpack-plugin';
```
Comment (igor-dv) [inline lib/ui/scripts/config.js]: I've used `terser-webpack-plugin` in my previous PR, and looks like webpack uses it by default now.

File: lib/ui/package.json [hunk 14]
```diff
@@ -20,9 +17,8 @@
   "main": "dist/index.js",
   "jsnext:main": "src/index.js",
   "scripts": {
-    "prepare": "node ../../scripts/prepare.js",
-    "publish-storybook": "bash ./.scripts/publish_storybook.sh",
-    "storybook": "start-storybook -p 9010"
+    "dll": "node -r esm ./scripts/script.js",
```
Comment (igor-dv) [inline lib/ui/package.json]: `scripts/script` ? `createDlls` ?

File: lib/ui/scripts/config.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  },
+  externals: provided,
+
+  module: {
+    rules: [
+      {
+        test: /\.jsx?$/,
+        use: [
+          {
+            loader: 'babel-loader',
```
Comment (igor-dv) [inline lib/ui/scripts/config.js]: look like there is no need for js rules since you are pointing entries to the `dist`

File: lib/ui/scripts/script.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
+import path from 'path';
+import webpack from 'webpack';
+
+import config from './config';
```
Comment (igor-dv) [inline lib/ui/scripts/script.js]: config => webpackDllsConfig

File: lib/ui/src/modules/ui/components/menu_item.stories.js [hunk 17]
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
-import { storiesOf } from '@storybook/react';
-import { action } from '@storybook/addon-actions';
+import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
```
Comment (igor-dv) [inline lib/ui/src/modules/ui/components/menu_item.stories.js]: what happened in all these files?

File: package.json [hunk 18]
```diff
@@ -121,7 +122,8 @@
     "tslint": "~5.11.0",
     "tslint-config-prettier": "^1.15.0",
     "tslint-plugin-prettier": "^2.0.1",
-    "typescript": "^3.1.6"
+    "typescript": "^3.1.6",
+    "uglifyjs-webpack-plugin": "^2.0.1"
```
Comment (igor-dv) [inline package.json]: let's use `terser-webpack-plugin`

**Review by igor-dv — DISMISSED** [review]
Shouldn't we "watch" on dlls?

> One of the techniques is a shared cache between the 2 instances of webpack

If manager will be moved to its own process, this option probably can't be used. Also, can't this cache become problematic with manager and preview having a different `react` versions?

File: lib/core/src/server/build-static.js (hunk 10 above)
Comment (igor-dv) [reply lib/core/src/server/build-static.js]: I got this one after reading the whole PR. So nvm.

File: lib/ui/src/modules/ui/components/menu_item.stories.js (hunk 17 above)
Comment (ndelangen) [reply lib/ui/src/modules/ui/components/menu_item.stories.js]: I wanted to remove the cyclical dependencies warnings from lerna

File: package.json (hunk 18 above)
Comment (ndelangen) [reply package.json]: sure!

File: lib/core/src/server/build-static.js (hunk 9 above)
Comment (ndelangen) [reply lib/core/src/server/build-static.js]: I couldn't find any usage of this though, how would this ever be used by users?

Comment (igor-dv) [reply lib/core/src/server/build-static.js]: Issue: #2762
PR: #2866

File: lib/core/package.json [hunk 19]
```diff
@@ -55,6 +55,7 @@
     "file-loader": "^2.0.0",
     "file-system-cache": "^1.0.5",
     "find-cache-dir": "^2.0.0",
+    "fs-extra": "^7.0.1",
```
Comment (igor-dv) [inline lib/core/package.json]: Oh, that's nice, wanted to do this in a separate PR, but  ❤️

File: lib/core/src/server/build-static.js (hunk 9 above)
Comment (ndelangen) [reply lib/core/src/server/build-static.js]: I see, I didn't know about this.

**Comment by ndelangen** [comment]
Some debugging to do:
```
Summary of all failing tests
 FAIL  examples/cra-kitchen-sink/src/storyshots.test.js
  ● Test suite failed to run

    Cannot find module 'App' from 'App.stories.js'

      7 | storiesOf('App', module).add('full app', () => <App />);
      8 |

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:221:17)
      at Object.<anonymous> (examples/cra-kitchen-sink/src/stories/App.stories.js:9:35)

 FAIL  examples/angular-cli/angularshots.test.js
  ● Test suite failed to run

    Cannot find module '/Users/dev/Projects/GitHub/storybook/core/examples/angular-cli/.storybook/config' from 'configure.js'

      36 |   const resolvedConfigPath = getConfigPathParts(configPath);
      37 |
    > 38 |   require.requireActual(resolvedConfigPath);
         |           ^
      39 | }
      40 |
      41 | var _default = configure;

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:221:17)
      at configure (addons/storyshots/storyshots-core/dist/frameworks/configure.js:38:11)
      at Object.load (addons/storyshots/storyshots-core/dist/frameworks/angular/loader.js:36:26)
      at loadFramework (addons/storyshots/storyshots-core/dist/frameworks/frameworkLoader.js:32:17)
      at testStorySnapshots (addons/storyshots/storyshots-core/dist/api/index.js:47:36)
      at Object.<anonymous> (examples/angular-cli/angularshots.test.js:22:30)


Test Suites: 2 failed, 112 passed, 114 total
```
